File: LeetCode/Easy/0225-implement-stack-using-queues/0225-implement-stack-using-queues.cs
public class MyStack {
    private readonly Queue<int> queue = new Queue<int>();
    
    public void Push(int x) {
        queue.Enqueue(x);
        for (var i = 0; i < queue.Count - 1; i++)
        {
            queue.Enqueue(queue.Dequeue());
        }
    }
    
    public int Pop() {
        return queue.Dequeue();
    }
    
    public int Top() {
        return queue.Peek();
    }
    
    public bool Empty() {
        return queue.Count == 0;   
    }
}


public class MyStack2Queues {
    private Queue<int> active;
    private Queue<int> aux;

    private int top;

    public MyStack2Queues() {
        active = new Queue<int>();
        aux = new Queue<int>();
    }
    
    public void Push(int x) {
        active.Enqueue(x);
        top = x;
    }
    
    public int Pop() {
        if (active.Count == 1)
        {
            top = default;
            return active.Dequeue();
        }

        while(active.Count > 2)
        {
            aux.Enqueue(active.Dequeue());
        }
        top = active.Peek();
        aux.Enqueue(active.Dequeue());
        var item = active.Dequeue();

        var temp = active;
        active = aux;
        aux = temp;

        return item;
    }
    
    public int Top() {
        return top;
    }
    
    public bool Empty() {
        return active.Count == 0;   
    }

    private void MoveAllButOne()
    {
        while(active.Count > 1)
        {
            aux.Enqueue(active.Dequeue());
        }
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack obj = new MyStack();
 * obj.Push(x);
 * int param_2 = obj.Pop();
 * int param_3 = obj.Top();
 * bool param_4 = obj.Empty();
 */

File: LeetCode/Easy/0225-implement-stack-using-queues/0225-implement-stack-using-queues.py
class MyStack:

    def __init__(self):
        self.queue = Deque()

    def push(self, x: int) -> None:
        self.top_x = x
        self.queue.append(x)


    def pop(self) -> int:
        self.top_x = None
        for _ in range(len(self.queue) - 1):
            self.top_x = self.queue.popleft()
            self.queue.append(self.top_x)
        
        return self.queue.popleft()


    def top(self) -> int:
        return self.top_x
        

    def empty(self) -> bool:
        return not self.queue

# 2-queues version - before follow-up
# class MyStack:
#     def __init__(self):
#         self.left = Deque()
#         self.right = Deque()
#         self.top_x = None

#     def push(self, x: int) -> None:
#         self.top_x = x
#         if self.left:
#             self.left.append(x)
#         else:
#             self.right.append(x)

#     def pop(self) -> int:
#         outbound, inbound = (self.left, self.right) if self.left else (self.right, self.left)
#         self.top_x = None
#         while len(outbound) > 1:
#             self.top_x = outbound.popleft()
#             inbound.append(self.top_x)

#         return outbound.popleft()

#     def top(self) -> int:
#         return self.top_x
        

#     def empty(self) -> bool:
#         return not self.left and not self.right


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()